diff --git a/bin/hint b/bin/hint
index bc6db7f..3d7fcec 100644
--- a/bin/hint
+++ b/bin/hint
@@ -127,7 +127,7 @@ def add_cnv_parser(subparsers):
     cnvparser.add_argument("--doiter",dest="doIterative",action='store_true',help = "If this switch is on, HiNT will do the iterative regression model by removing copy numer variated regions, DEFAULT=False",required=False,default=False)
     cnvparser.add_argument("-f","--format",dest="format",type=str,required = False, choices = ("cooler","juicer"), default="cooler",
                          help="Format for the output contact matrix, DEFAULT: cooler")
-    cnvparser.add_argument("-e","--enzyme",dest="enzyme",choices = ("MboI","HindIII","DpnII"), default="MboI",required = False, help="enzyme used for the Hi-C experiments, will be used to calculate enzyme sites")
+    cnvparser.add_argument("-e","--enzyme",dest="enzyme",choices = ("arima","MboI","HindIII","DpnII"), default="MboI",required = True, help="enzyme used for the Hi-C experiments, will be used to calculate enzyme sites")
     cnvparser.add_argument("-r","--resolution",dest="resolution",type=int,required = False, default=50,
                          help="Resolution for the Hi-C contact matrix used for the CNV detection, unit: kb, DEFAULT: 50kb")
     cnvparser.add_argument("-g","--genome",dest="genome", choices = ("hg38","hg19","mm10"), default="hg19",required=False,
@@ -152,7 +152,7 @@ def add_translocation_parser(subparsers):
                          or the directory that contain Hi-C interaction matrix in sparse or dense matrix format, interchromosomal interaction matrices only. Absolute path is required")
     tranlparser.add_argument("--refdir",dest="referencedir",type=str,required = True,
                          help="the reference directory that downloaded from dropbox dropbox. (https://www.dropbox.com/sh/2ufsyu4wvrboxxp/AABk5-_Fwy7jdM_t0vIsgYf4a?dl=0.)")
-    tranlparser.add_argument("-e","--enzyme",dest="enzyme",type=str,required = False, choices = ("DpnII","MboI","HindIII"), default="MboI",
+    tranlparser.add_argument("-e","--enzyme",dest="enzyme",type=str,required = True, choices = ("arima", "DpnII","MboI","HindIII"), default="MboI",
                          help="Enzyme used in Hi-C experiment, DEFAULT: MboI")
     tranlparser.add_argument("-f","--format",dest="format",type=str,required = False, choices = ("cooler","juicer"), default="cooler",
                          help="Format for the output contact matrix, DEFAULT: cooler")
